internal/impl/azure/output_adls.go

```diff
@@ -0,0 +1,158 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azure
+
+import (
+	"context"
+	"fmt"
+
+	dlservice "github.com/Azure/azure-sdk-for-go/sdk/storage/azdatalake/service"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	// Azure Data Lake Storage Output Fields
+	adlsFieldFilesystem = "filesystem"
+	adlsFieldPath       = "path"
+)
+
+type adlsConfig struct {
+	client     *dlservice.Client
+	path       *service.InterpolatedString
+	filesystem *service.InterpolatedString
+}
+
+func init() {
+	err := service.RegisterOutput("azure_data_lake_gen2", adlsSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (out service.Output, mif int, err error) {
+			var pConf *adlsConfig
+			if pConf, err = adlsConfigFromParsed(conf); err != nil {
+				return
+			}
+			if mif, err = conf.FieldMaxInFlight(); err != nil {
+				return
+			}
+			if out, err = newAzureADLSWriter(pConf, mgr.Logger()); err != nil {
+				return
+			}
+			return
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func adlsConfigFromParsed(pConf *service.ParsedConfig) (*adlsConfig, error) {
+	var conf adlsConfig
+	var err error
+	conf.filesystem, err = pConf.FieldInterpolatedString(adlsFieldFilesystem)
+	if err != nil {
+		return nil, err
+	}
+	conf.path, err = pConf.FieldInterpolatedString(adlsFieldPath)
+	if err != nil {
+		return nil, err
+	}
+	var isFilesystemSASToken bool
+	conf.client, isFilesystemSASToken, err = adlsClientFromParsed(pConf, conf.filesystem)
+	if err != nil {
+		return nil, err
+	}
+	if isFilesystemSASToken {
+		conf.filesystem, _ = service.NewInterpolatedString("")
+	}
+	return &conf, nil
+}
+
+func newAzureADLSWriter(conf *adlsConfig, log *service.Logger) (*azureADLSWriter, error) {
+	return &azureADLSWriter{
+		conf: conf,
+		log:  log,
+	}, nil
+}
+
+type azureADLSWriter struct {
+	conf *adlsConfig
+	log  *service.Logger
+}
+
+func (a *azureADLSWriter) Connect(ctx context.Context) error {
+	return nil
+}
+
+func (a *azureADLSWriter) Write(ctx context.Context, msg *service.Message) error {
+	fsName, err := a.conf.filesystem.TryString(msg)
+	if err != nil {
+		return fmt.Errorf("interpolating filesystem name: %w", err)
+	}
+	path, err := a.conf.path.TryString(msg)
+	if err != nil {
+		return fmt.Errorf("interpolating file path: %w", err)
+	}
+	mBytes, err := msg.AsBytes()
+	if err != nil {
+		return fmt.Errorf("reading message body: %w", err)
+	}
+
+	fileClient := a.conf.client.NewFileSystemClient(fsName).NewFileClient(path)
+	_, err = fileClient.Create(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("creating file: %w", err)
+	}
+	err = fileClient.UploadBuffer(ctx, mBytes, nil)
+	if err != nil {
+		return fmt.Errorf("uploading message body: %w", err)
+	}
+	return nil
+}
+
+func (a *azureADLSWriter) Close(ctx context.Context) error {
+	return nil
+}
+
+func adlsSpec() *service.ConfigSpec {
```

Comment:
Small nit: We usually place the spec towards the upper part of the file, but no worries if you wish to deviate from that pattern.